mesonbuild/build.py

```diff
@@ -2359,10 +2359,22 @@ def get_default_object(self) -> BuildTarget:
 
 class CommandBase:
 
-    depend_files: T.List[File]
+    depend_files: T.List[FileOrString]
     dependencies: T.List[T.Union[BuildTarget, 'CustomTarget']]
     subproject: str
 
+    def process_depends(self,
+                        depend_files: T.Optional[T.Sequence[FileOrString]],
+                        extra_depends: T.Optional[T.Sequence[T.Union[BuildTarget, CustomTarget, str, File]]]) -> None:
+        self.depend_files = list(depend_files or [])
+        self.dependencies: T.List[T.Union[BuildTarget, CustomTarget]] = []
+        for d in extra_depends:
+            if isinstance(d, (BuildTarget, CustomTarget)):
+                self.dependencies.append(d)
+            elif isinstance(d, (str, File)):
+                FeatureNew.single_use('String or File in depends keyword argument', '1.0.0', self.subproject)
+                self.depend_files.append(d)
```

Comment:
Please no. CommandBase is a typing nightmare and It's going away in my build.py refactors, absolutely no adding to it please. Also, the build layer is the wrong place for any FeatureNew stuff to be happening, it needs to be done in the interpreter. Preferably the type splitting should happen there as well, and the build targets should be passed the values pre-separated.

You might be interested in the first couple of patches of my build.py stuff, which adds a feature validator callable to the `TypedKwarg`, which would be handy for this case.

Reply:
> Please no. CommandBase is a typing nightmare and It's going away in my build.py refactors, absolutely no adding to it please.

Wondering what can possibly be complicated for type checking in such trivial base class. The reason I added that code there is because it's where we define depend_files and dependencies, keeps common code together. Those 2 attributes are used by flatten_command() in the same base class. If your refactor lands first I'll adapt.

> Preferably the type splitting should happen there as well, and the build targets should be passed the values pre-separated.

I don't think it should be pre-separated because
- It would have to be done in every places we create a CustomTarget, many modules does it.
- The split is not an interpreter thing, but really just convenience for the backend. Doing it in build.py is the right place IMHO.

> Also, the build layer is the wrong place for any FeatureNew stuff to be happening, it needs to be done in the interpreter

I learned today that typed_kwarg can do that with `since_values`, I added a commit to extend it a bit to be able to do this there now.